tasks/shell_crmsh/crm-cib-constraint-location.yml

```diff
@@ -0,0 +1,50 @@
+# SPDX-License-Identifier: MIT
+---
+- name: Define location constraint.id {{ constraint.id | d(constraint_index) }}
+  ansible.builtin.set_fact:
+    __ha_cluster_constraint_id:
+      "{{ constraint.id if constraint.id is defined
+       else (constraint.resource.pattern | regex_replace('[^A-Za-z0-9]', '') | quote) + '-location'
+       if constraint.resource.pattern is defined else (constraint.resource.id | quote) + '-location' }}"
+
+# Verify if Shadow CIB already contains same constraint id.
+- name: Verify location constraint presence {{ __ha_cluster_constraint_id }}
+  ansible.builtin.command:
+    cmd: crm -c {{ __ha_cluster_crm_shadow }} configure show {{ __ha_cluster_constraint_id }}
+  register: __ha_cluster_constraint_status
+  changed_when: false
+  failed_when: false
+
+# Delete constraint id in Shadow CIB to avoid duplicate errors during cibadmin patch.
+- name: Delete present location constraint {{ __ha_cluster_constraint_id }}
+  ansible.builtin.command:
+    cmd: crm -c {{ __ha_cluster_crm_shadow }} configure delete {{ __ha_cluster_constraint_id }}
+  when: __ha_cluster_constraint_status.rc == 0
+  check_mode: false
+  changed_when: not ansible_check_mode
+
+- name: Configure location constraint {{ __ha_cluster_constraint_id }}
+  ansible.builtin.command:
+    cmd: |
```

Comment:
note that for complex templates, you can define these in a templates/ file and use the `template` lookup to convert the template to a string that you can pass to `command`, `copy.content`, etc.
this is fine, but fyi for future - https://docs.ansible.com/ansible/latest/collections/ansible/builtin/template_lookup.html#ansible-collections-ansible-builtin-template-lookup

Reply:
Looks like another structure equality between the shell_pcs and shell_crmsh subdirectories and code?

@tomjelinek should weigh in here too, as this might be tech debt to address across the whole Ansible Role dating back to Ansible Core 2.9 compatibility (when template lookup was very different) ?

Ref: https://github.com/linux-system-roles/ha_cluster/blob/1.13.5/tasks/shell_pcs/pcs-cib-constraint-location.yml#L3-L6

Reply:
Sure - just FYI - most devs don't know about the `template` lookup, and think they can only use template files with the `template` module.

Reply:
My intent was structure equality as Sean pointed out, plus I wanted avoid creation of temporary files with template module, so I opted to follow same as you did with pcs inline jinja2.

It was not easy, but current j2 code is generating crmsh compliant commands and creates resources same as pcs.

Reply:
I didn't even know it was possible to use template files in this way. Still, I like the inline templates more for this purpose, it reduces the number of files in the role and thus it makes maintenance a bit easier. Also, if the command templates were separate files, one could think that they are used as a basis for files produced by the role, which is not true.

Reply:
> I didn't even know it was possible to use template files in this way. Still, I like the inline templates more for this purpose, it reduces the number of files in the role and thus it makes maintenance a bit easier.

Ok.

> Also, if the command templates were separate files, one could think that they are used as a basis for files produced by the role, which is not true.

Ok.

Reply:
Decision made, @richm please mark as resolved if you agree.